### PR TITLE
Preserve lifted SymInts during strict export capture

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185082

Strict export skipped Dynamo basic-symbol lifting while creating graph inputs. This made the export torch-level graph diverge from Dynamo: tensor shape symbols were recomputed with `aten.sym_size` during later tracing instead of being consumed as lifted `SymInt` placeholders.

Enable input-symbol lifting only for strict export's bytecode capture path, which preserves `TensorPropertySource` metadata. Keep internal AOT inputs separate from the public input signature, run symbolic CSE without emitting export runtime asserts, and unlift source-marked tensor properties to `sym_size`, `sym_stride`, or `sym_storage_offset` only at the public `ExportedProgram` boundary. Public inputs pruned by AOT are reintroduced from their tracked source metadata, including constants, module state, HOP operands, and shape-only graph inputs.

Follow-up CI exposed public-source and calling-convention gaps. Export now preserves explicit, variadic, and non-literal dictionary-key Dynamo sources; canonicalizes ambiguous internal/non-literal guard sources to serialized flat-argument sentinels; and resolves those sentinels only when unlifting the program. Guard-source substitutions run longest-first so overlapping source strings cannot corrupt each other, and absent root values are handled without dereferencing the sentinel. AOT bytecode calls recover flattened symbolic inputs and pytree specs, while AOTI snapshots and exception-safely restores original parameters, buffers, and non-persistent-buffer registration after functionalization.

The alternative was to keep export's early `sym_size` behavior and patch later graphs or expectations. That would retain the Dynamo/export divergence and continue losing source identity for variadic, pruned, or non-literal-key inputs, so the fix keeps the internal graph lifted and makes the public unlift boundary explicit.

Known limitation: `test_export_preserves_unused_tensor_with_enum_dict_key` remains an expected failure in the generated SerDes, SerDes non-strict, TrainingIR-to-run-decomposition, and non-strict TrainingIR-to-run-decomposition variants because those downstream pipelines do not yet support the local-enum/tensor-key combination.

Fixes #165073

Generated by my agent

Test Plan:
- `python -m pytest -q` focused `TestDynamismExpression` lift/unlift, flattened-AOT, source-matching, dict-key, guard-serialization, kwargs, strict-v2, SerDes, training-IR, unflatten, and cond tests
- `python -m pytest -q test/export/test_export.py::TestExport::test_guards_fn_recovers_from_unparse_recursion_error test/export/test_export.py::TestExport::test_guards_fn_replaces_longest_source_first test/export/test_export.py::TestExport::test_dynamic_shapes_nested_kwargs_overlapping_source_names test/export/test_retraceability.py::RetraceExportNonStrictTestExport::test_dynamic_shapes_nested_kwargs_overlapping_source_names_retraceability_nonstrict`
- `python -m pytest -q test/export/test_retraceability.py::RetraceExportTestExport::test_dynamic_shapes_nested_kwargs_overlapping_source_names_retraceability_strict`
- `python test/export/test_converter.py TestConverter.test_aten_len -v`
- `python test/inductor/test_aot_inductor.py TestAOTInductorConfig.test_preserve_module_parameters_and_buffers_on_error -v`
- `python -m pytest -q test/export/test_strict_export_v2.py -k 'export_cond'`
- `python -m py_compile` on all changed Python files
- `lintrunner -a`
- `git diff --check`
- `BASE=main HEAD=HEAD bash .github/scripts/pr-sanity-check.sh`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR and addressed the unclassified export CI regressions: strict baddbmm now preserves get_attr targets during bytecode capture, and strict retraceability stays on the existing GraphModule capture path. I also added direct coverage for stride/storage-offset unlift handling and reran the focused export tests plus lintrunner.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and pushed follow-up fixes for the review and CI findings. The cleanup removes confusing placeholder-boundary logic, the AOTInductor runtime-check failure is addressed by handling AOTAutograd's already-flattened bytecode-codegen calls, and the serialization failures looked like resource/OOM kills rather than PR-caused failures.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR on current main and fixed the PR-caused CI buckets from the latest Dr. CI run. The strict export unlift step now reintroduces public tensor placeholders pruned by AOT, bytecode capture runs export constraint solving before detaching fake mode, install_free_tensors falls back to the compatible capture path, and the runtime_assert public signature stays backcompat-safe.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the latest PR-caused Dr. CI export failures. The extra post-capture constraint solver now only runs for bytecode-captured export graphs, and its equality sources use the same root-input source marking as Dynamo so reused Dim constraints are checked against the captured ShapeEnv. I also updated the strict vmap expected graph for the resulting stable sym_size name.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the latest PR-caused export CI failures. The bytecode export constraint solver now runs inside the capture helper on Dynamo's captured ShapeEnv instead of mixing in equality metadata from a fresh fake-input pass, and the public graph drops the temporary tracing context so deepcopy-sensitive export paths work again. I also removed the stale strict-export-v2 xfail for constant input naming.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the latest strict export/AOTInductor CI failures. The export boundary now maps public metadata by signature names, keeps lifted module parameter/buffer sources classified as non-user inputs, removes duplicate internal tensor placeholders before public signature/range-constraint generation, and drops the stale strict-v2 expected failure. I reran the focused export regressions plus lintrunner; the full AOTInductor tests still cannot import locally because torchvision is built for CUDA 13 while this checkout is CUDA 12.8.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the latest Dr. CI failures. Dynamic-shape constraint generation now ignores specialized constant leaves when aligning specs with public placeholders, strict export reintroduces missing public constant inputs so the callable signature stays valid, HOO subgraph placeholder naming handles duplicate lifted SymInt names, and get_attr targets are restored before strict export parameter/buffer metadata collection. I also tightened duplicate tensor placeholder cleanup so it requires tensor identity or matching tracked fake source instead of shape metadata alone. Lint and py_compile passed; the focused strict export test hung in this local host, and AOTInductor still cannot import locally because torchvision is CUDA 13 while this build is CUDA 12.8.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR and fixed the latest PR-caused CI failures. get_attr target restoration now treats FX's missing-attribute AssertionError as non-fatal in both export and Dynamo bytecode capture paths, and strict export's unlift/dedup logic now matches rebuilt public sources by stable source name so Wrapped.forward(*args) inputs are not duplicated or collapsed into the wrong tensor. I reran the focused strict export, retrace, serdes, exportdb, serialize, joint-basic, and SymInt unlift tests plus lintrunner; local test_fx and AOTInductor imports are still blocked by the host torchvision CUDA 13 vs PyTorch CUDA 12.8 mismatch.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and updated the PR to fix the latest strict export CI failures. The new changes prune lifted tensor-property SymInt operands from supported HOP input lists without changing while_loop carried-state arity, restore dynamic tensor-constant get_attr targets before metadata collection, keep ShapesSpec on the transformer/runtime-assert path, and remove the stale strict-v2 enum expected failure. I reran lintrunner -a, git diff --check, git diff --cached --check, focused cond/serdes, dynamic spec runtime-assert, strict-v2 enum, dynamic tensor-constant, and standalone while_loop schema/runtime repros. The full AOTInductor and functorch test files still cannot import locally because this host has torchvision CUDA 13 with PyTorch CUDA 12.8.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and pushed fixes for the new Dr. CI failures: strict/export inline expectation drift, the strict-v2 expected-failure xpass, and the AOTI unlift crash where aot_export_module left FunctionalTensor-wrapped module attrs. Local focused export tests, a direct nested_tensor_from_jagged AOTI compile repro, and lintrunner -a passed; the new CI run is now pending.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated this PR to address the new strict export and AOTI CI failures. The strict export path now re-fakifies unused public tensor inputs before source collection so reintroduced placeholders preserve the requested dynamic/static shape specs, removed the stale strict-v2 expected-failure marker, and AOTI now serializes call specs for bytecode export graphs so nested_tensor_from_jagged loads no longer fail with an empty JSON call spec.

**Comment by jansel** [comment]
My agent says I rebased and pushed fixes for the latest Dr. CI failures. The strict export shape-only path now avoids re-fakifying under a fake mode that cannot track new fakes and instead relies on TensorPropertySource recovery from the lifted symbol ShapeEnv. The AOTI path now restores original module parameters and buffers after aot_export_module in a finally block, preserving non-persistent buffer registration, so nested_tensor_from_jagged AOTI compile/load no longer leaves the original module with FunctionalTensor attrs.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I pushed a CI fix for the strict-export-v2 failures: test_if_post_autograd_op_preserved now passes under strict v2, so I removed its stale expectedFailureStrictV2 marker. I reran the base test, the generated strict-v2 test, lintrunner -a, and diff checks. The remaining Dr. CI inductor_timm bucket reported tf_efficientnet_b0 as eager_two_runs_differ, which points to eager nondeterminism rather than a compiled-output regression.

**Comment by jansel** [comment]
My agent says I rebased the PR and pushed fixes for the latest Dr. CI export/control-flow failures. The generated strict-export-v2 nested-module tests now pass, so I removed the stale expectedFailureStrictV2 markers, and the strict dynamic while_loop export expected graph now reflects the lifted-symbol Dim name that the PR intentionally preserves. I reran the targeted strict-v2 nested-module tests, the control-flow repro, the related export strict check, lintrunner -a, and diff checks. The remaining inductor TorchBench/TIMM buckets reported eager_two_runs_differ/eager failures, which point to eager nondeterminism rather than a compiled-output regression.

**Comment by jansel** [comment]
My agent says I rebased and pushed PR 185082 for the latest Dr. CI failures. The strict-export-v2 no_tensor_computation_2 case now passes, so I removed its stale expectedFailureStrictV2 marker. A required review pass also found a requires_grad metadata edge case where string constant user inputs could collide with tensor placeholder names; I fixed _preserve_requires_grad_pass to derive tensor input names from input_specs and added coverage for that collision across base, strict, and strict-export-v2 export variants. The reported dtypeview and dynamic_shapes CI buckets pass locally with the repo env libstdc++ and a temporary empty torchvision import shim; the CI logs I inspected showed 30-minute timeouts rather than assertion failures. I reran the focused export checks, the dtypeview and dynamic-shapes repros, lintrunner -a, and diff checks.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the repeated strict-export-v2 CI failure. test_placeholder_naming_collisions_hoo_subgraphs now passes under strict export v2, so I removed its stale expectedFailureStrictV2 marker and reran the generated strict-v2 test, the base test, focused lifted-SymInt export tests, diff checks, and lintrunner -a. I also probed the linalg_polar opinfo bucket locally; it reproduces as Dynamo's existing unsupported linalg_polar baseline and Dr. CI also classified the same op as flaky/broken-trunk in other shards.

**Comment by jansel** [comment]
My agent says I fixed the latest Dr. CI strict-export-v2 failures. test_placeholder_naming_collisions_strict_export_v2 now passes, so I removed the stale expectedFailureStrictV2 marker from the base placeholder naming collision test. I reran the generated strict-v2 test, the base test, the neighboring strict-v2 HOO collision test, lintrunner -a, and diff checks.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the latest Dr. CI strict-export-v2 XPASS failures. test_placeholder_naming_order and test_placeholder_naming_order_variadic now pass under strict export v2, so I removed their stale expectedFailureStrictV2 markers. I reran the generated strict-v2 pair, the base tests, py_compile for the conflicted export helper and test file, lintrunner -a, and diff checks. The distributed job failure in Dr. CI failed before tests ran in runner-container-hooks with a null jobPod error, so there is no PR code change for that bucket.

**Comment by jansel** [comment]
My agent says I pushed a CI fix for the six Dr. CI failures in test_preserve_shape_dynamism_for_unused_inputs_strict_export_v2. The generated strict-export-v2 test now passes, so I removed the stale expectedFailureStrictV2 marker from the base preserve_shape_dynamism_for_unused_inputs test. I reran the generated strict-v2 test, the base test, lintrunner -a, and diff checks.

**Comment by jansel** [comment]
My agent says I fixed the latest Dr. CI strict-export-v2 XPASS failures for test_run_decompositions_keep_metadata_strict_export_v2. The generated strict-export-v2 test now passes, so I removed the stale expectedFailureStrictV2 marker from the base test and reran the generated strict-v2 test, the base test, lintrunner -a, and diff checks.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says Dr. CI was failing the dynamic_shapes_kwargs_call_like generated export variants because public input source collection rebinding raw kwargs made variadic kwargs look like standalone locals. I updated export to preserve the original explicit/*args/**kwargs Dynamo sources, map guard source names to the public ExportedProgram signature with a longest-prefix match, and added targeted kwargs collision/shared-dim coverage.

**Comment by jansel** [comment]
My agent says I pushed fixes for the latest Dr. CI blockers. The strict-export-v2 tensor-constant metadata test now passes, so I removed its stale expectedFailureStrictV2 marker; I also consolidated adjacent regression tests while preserving the size/stride/storage-offset, explicit SymInt, kwargs source, and requires_grad name-collision coverage so pr-sanity-checks now reports PR SIZE 1999 locally. Focused base and strict-export-v2 tests plus lintrunner passed.

**Comment by jansel** [comment]
My agent says I fixed the latest Dr. CI strict-export-v2 failures for test_sdpa_gqa_strict_export_v2. The generated strict export v2 test now passes, so I removed the stale expectedFailureStrictV2 marker from the base test and reran the base/generated pair, touched-file lint, diff checks, and pr-sanity-checks.

**Comment by jansel** [comment]
My agent says I updated the PR for the Dr. CI export failures. The converter failure was caused by public input source collection treating tensor/enum dict keys as literal DictGetItemSource keys; the fix now preserves non-literal dict-key sources by stable dict position. I also fixed guard-source rewriting so internal flat_args guard sources cannot collide with a real user flat_args argument and survive export save/load.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased on current main and addressed the latest review and CI failures. Guard sources are now canonicalized once at capture to serialized sentinels for internal, non-literal, and ambiguous flat_args inputs, so save/load remains unambiguous; the invariant public-source ordering is computed once and the serialized-token contract is documented. The unsupported SerDes/training-IR dict-key variants are marked expected, the stale strict-v2 unflatten XFAIL is removed, HOP unlift no longer leaves dead tensor-property nodes, and AOTI state restoration now has exception/non-persistent-buffer coverage. Focused export/generated tests, the rebased cond group, full lintrunner, and pr-sanity-checks pass at PR SIZE 1975.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the six new CI failures by removing the stale non-strict retraceability expected-failure marker now that the nested kwargs case passes. I also addressed the latest review feedback: guard-source substitutions now run longest-first, missing root values no longer dereference the sentinel, and focused regression coverage exercises overlapping source strings. The targeted export/retraceability tests, lintrunner, diff checks, and the 1,996-line PR sanity check pass.

**Comment by jansel** [comment]
My agent says I refreshed the PR description to match the consolidated current test plan and explicitly documented the expected-failure SerDes and training-IR variants noted in the latest automated review.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
